### PR TITLE
Fix the sgen issue when use NaN as the default value. 

### LINKED ISSUE #19724: The generated code cannot compile if define the class that use NaN as the default value
If to generate the code for the classes that contains the following, the generated code cannot compile. This also repro on the desktop version.

```
[DefaultValue(double.NaN)]
public double DoubleProp { get; set; }
```

The generated code is like. Cannot recognize NaN.
`if (((global::System.Double)o.@DoubleField) != NaN) `

### PR DESCRIPTION
**Pull request by huanwu**
Fix the sgen issue when use NaN as the default value. 
Move the NaN test from XmlSerializerTest .runtimeonly to XmlSerializerTest so it will be included in Sgen test run.

Fix #19724

@shmao @zhenlan @mconnew

### CODE REVIEW

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationWriter.cs [hunk 1]
```diff
@@ -2044,7 +2044,14 @@ private void WriteValue(object value)
                 else if (type == typeof(Int32))
                     Writer.Write(((Int32)value).ToString(null, NumberFormatInfo.InvariantInfo));
                 else if (type == typeof(Double))
-                    Writer.Write(((Double)value).ToString("R", NumberFormatInfo.InvariantInfo));
+                    if (double.IsNaN((Double)value))
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationWriter.cs]: Add brackets around the `if...else...`

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationWriter.cs [hunk 2]
```diff
@@ -2057,10 +2064,15 @@ private void WriteValue(object value)
                     Writer.Write(")");
                 }
                 else if (type == typeof(Single))
-                {
-                    Writer.Write(((Single)value).ToString("R", NumberFormatInfo.InvariantInfo));
-                    Writer.Write("f");
-                }
+                    if (Single.IsNaN((Single)value))
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationWriter.cs]: Add brackets around the `if...else...`

**Comment by huanwu** [comment]
@dotnet-bot test NETFX x86 Release Build

**Comment by huanwu** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by huanwu** [comment]
@dotnet-bot test UWP CoreCLR x64 Debug Build

**Comment by huanwu** [comment]
@dotnet-bot test UWP NETNative x86 Release Build

**Comment by huanwu** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by huanwu** [comment]
@dotnet-bot test Windows x64 Debug Build

**Comment by huanwu** [comment]
@dotnet-bot test Windows x86 Release Build

**Comment by huanwu** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by huanwu** [comment]
@dotnet-bot test UWP CoreCLR x64 Debug Build

**Comment by huanwu** [comment]
Unrelated tests ReceiveFromV6BoundToSpecificV6_Success and UnloadingEventMustHappenBeforeProcessExitEvent fail in Linux and UWP. Unrelated to my change.

**Comment by huanwu** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by huanwu** [comment]
@dotnet-bot test UWP CoreCLR x64 Debug Build

**Pull request merged**